### ISSUE TITLE
Exposing the relations between register names

### ISSUE THREAD

**Issue by meithecatte** [iximeow/yaxpeax-x86#30]
I find myself writing some ugly code to answer questions like "which 64-bit GPR is `ch` a part of? which bits?".

I would expect `yaxpeax` to provide APIs on `RegSpec` for this, though I understand if this is something you want to model in `yaxpeax_arch`, with the API design questions this involves.

**Comment by iximeow** [comment]
agreed on all counts! [`yaxpeax_core`](https://github.com/iximeow/yaxpeax-core/blob/87786db784bff488ad374b751b1ccc1bd75036eb/src/data/mod.rs#L18-L45) has a pretty unsatisfactory interface for this that i'd never gotten happy enough with to move over to `arch`, but i definitely agree that "this register selects the u64 you're using to represent `rcx`" or similar should be something this crate can answer. reading between the lines, that's kind of related to what @i509VCB was getting at over [here](https://github.com/iximeow/yaxpeax-x86/issues/11) (agh goodness that's an old issue)

so, that said, i think something more specifically suited to x86 is a great excuse to think about what a good interface would feel like.

things off the top of my head:
* "what is the largest alias of this register?" (`ch -> rcx`)
* "what is the smallest larger alias of this register?" (`ch -> cx`, `cx -> ecx`, `ecx -> rcx`)
* "what registers are aliased by the largest alias of this register?" (`ch -> {rcx, ecx, cx, cl}`, maybe include `ch` in the list idk)
* "what registers are aliased by the smallest larger alias of this register?" (`ch -> {cx, cl}`, maybe include `ch` in the list idk)
* bonus: "where is this register, in that other register?" (`ch -> (rcx, 8)`, `cl -> (rcx, 0)`)

and of course at the minimum, "what is the largest alias?" and "what is the smallest larger alias?" should be `const fn`.

**Comment by iximeow** [comment]
.... anyway another two years on, i kind of have an idea of how this might work out: https://github.com/iximeow/yaxpeax-x86/issues/11#issuecomment-3441763934

in `yaxpeax_arch` i think there's probably a reasonable
```
trait Arch {
    type Register: RegisterGeometry;
}

trait RegisterGeometry: Eq {
    fn offset_in(&self, other: &Self) -> Option<u64>;
    fn width(&self) -> u64;
    fn contains(&self, other: &Self) -> bool;
    fn overlaps(&self, other: &Self) -> bool;
    fn next_largest_enclosing(&self) -> Self;
    fn as_largest_enclosing(&self) -> Self {
        let mut reg = self.next_largest_enclosing();
        while reg.next_largest_enclosing() != reg {
            reg = reg.next_largest_enclosing();
        }
        reg
    }
}
```

which could happen to have efficient impls (on both x86 and a64!)

**Comment by meithecatte** [comment]
Hm, so that trait lets you go "up", but does it let you go "down"? (ah, looks like that'd get handled by the part of the design explained in the comment you linked)

**Comment by iximeow** [comment]
yeah, i don't think there's a nice abstract mechanism for picking smaller sizes like that. two years ago i was concerned about a case like, what if a register existed that was like
```
| eax[3] | eax[2] | eax[1] | eax[0] | ebx[3] | ebx[2] | ebx[1] | ebx[0] |
                  |       <evil register here>        |
```

and _that_ i'm willing to say is unreasonable at this point. if an architecture really had to represent that, `Arch::Register` probably would "just" need to have a representation of the register pair or something. a concept for whatever the next largest whole "thing" is.

(i _can_ imagine such a mechanism, i'm just not sure it's "nice"! if each register class had a distinct type and there was a generic `Register::reinterpret_class<Bank: RegisterBank>()` you could write `RegSpec::rsp().reinterpret_class::<RegisterBank::B>()`, returning `Some(RegSpec::spl())` for example. but _there_ i'm confident about rigid distinctions between kinds of registers like that. it could be reasonably added to `RegisterGeometry` in the future though)
